### PR TITLE
Tried to add tail call optimization

### PR DESCRIPTION
**Pull request by mistasse**

### CODE REVIEW

File: src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		final Object name = labelNode.executeLabel(message);
+		assert OzGuards.isLiteral(name);
+
+		OzProc method = getMethod(self, name);
+		if (method == null) { // redirect to otherwise
+			method = getMethod(self, "otherwise");
+			message = OTHERWISE_MESSAGE_FACTORY.newRecord(message);
+		}
+
+		throw new TailCallException(method, new Object[] { self, message });
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java]: This logic should be extracted to a different node, like `CallMethodNode`.

File: src/org/mozartoz/truffle/runtime/TailCallException.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package org.mozartoz.truffle.runtime;
+
+
+public class TailCallException extends RuntimeException {
```
Comment (eregon) [inline src/org/mozartoz/truffle/runtime/TailCallException.java]: Extend ControlFlowException, so this allows the optimization in Graal.

File: src/org/mozartoz/truffle/translator/Translator.java [hunk 3]
```diff
@@ -181,6 +185,21 @@ public OzRootNode translateAST(String description, Statement ast, Function<OzNod
 	}
 
 	OzNode translate(StatOrExpr node) {
+		if(node instanceof TailMarkerStatement) {
```
Comment (eregon) [inline src/org/mozartoz/truffle/translator/Translator.java]: missing space after the `if`, normally the Eclipse formatter should take care of this.

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import org.mozartoz.truffle.runtime.OzArguments;
+import org.mozartoz.truffle.runtime.TailCallException;
+
+import com.oracle.truffle.api.dsl.CreateCast;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+
+public class TailCallCatcherNode extends OzNode {
+
+	@Child OzNode call;
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: You can type this as a CallProcNode

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		TailCallException tailCall = null;
+		while (true) {
+			try {
+				if (tailCall == null) {
+					return call.execute(frame);
+				} else {
+					return callNode.call(frame,
+							tailCall.proc.callTarget,
+							OzArguments.pack(tailCall.proc.declarationFrame, tailCall.arguments)
+							);
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: And then reuse the CallProcNode here, with something like
`executeCall(proc, arguments)`.
But CallProcNode needs to be adapted to allow taking direct arguments and not execute the arg nodes.

File: src/org/mozartoz/truffle/translator/Translator.java [hunk 6]
```diff
@@ -181,6 +185,21 @@ public OzRootNode translateAST(String description, Statement ast, Function<OzNod
...
+			//return translate(((TailMarkerStatement) node).call());
+
+			CallStatement call = ((TailMarkerStatement) node).call();
+			List<Expression> args = new ArrayList<>(toJava(call.args()));
+			OzNode[] argNodes = new OzNode[args.size()];
+			for (int i = 0; i < argNodes.length; i++) {
+				argNodes[i] = translate(args.get(i));
+			}
```
Comment (eregon) [inline src/org/mozartoz/truffle/translator/Translator.java]: A bit of duplicated logic here.
We should probably have a `ExecuteValuesNode` that just execute a OzNode[] and returns a Object[].

**Review by eregon — CHANGES_REQUESTED** [review]
Looks good overall but it needs a few nicer refactorings :)

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	@CreateCast("function")
+	protected OzNode derefFunction(OzNode var) {
+		return DerefNode.create(var);
+	}
+
+	@Override
+	public Object execute(VirtualFrame frame) {
+		TailCallException tailCall = null;
+		while (true) {
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: There should be some profiling here, to make sure normal calls don't get overhead.

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java (hunk 5 above)
Comment (eregon) [reply src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: > But CallProcNode needs to be adapted to allow taking direct arguments and not execute the arg nodes.

Latest master allows just that.

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+public class TailCallCatcherNode extends OzNode {
+
+	@Child CallNode callNode;
+
+	public TailCallCatcherNode(CallNode callNode) {
+		this.callNode = callNode;
+	}
+
+	BranchProfile tailCallProfile = BranchProfile.create();
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: Profiles must be `final` (and usually are `private` as well)

File: src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	public Object execute(VirtualFrame frame) {
+		try {
+			return callNode.execute(frame);
+		} catch (TailCallException e) {
+			tailCallProfile.enter();
+			while (true) {
+				try {
+					return callNode.executeCall(frame, e.receiver, e.arguments);
+				} catch (TailCallException exception) {
+					e = exception;
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java]: Reassigning a catch variable is not too nice, could you make a tailCall local var?

File: src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+@NodeChildren({ @NodeChild("receiver"), @NodeChild("arguments") })
+public abstract class TailCallThrowerNode extends OzNode {
+
+	@CreateCast("receiver")
+	protected OzNode derefReceiver(OzNode var) {
+		return DerefNode.create(var);
+	}
+
+	public abstract Object executeTailCall(VirtualFrame frame, Object receiver, Object[] arguments);
```
Comment (eregon) [inline src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java]: I think this is not needed.

File: src/org/mozartoz/truffle/runtime/TailCallException.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package org.mozartoz.truffle.runtime;
+
+import com.oracle.truffle.api.nodes.ControlFlowException;
+
+public class TailCallException extends ControlFlowException {
+
+	private static final long serialVersionUID = -6432472909387052509L;
+
+	public final Object receiver;
+	public final Object[] arguments;
```
Comment (eregon) [inline src/org/mozartoz/truffle/runtime/TailCallException.java]: Use `@CompilationFinal(dimensions=1)` here, it allows to fold reading the elements.

File: src/org/mozartoz/truffle/translator/Translator.java [hunk 12]
```diff
@@ -272,7 +277,9 @@ OzNode translate(StatOrExpr node) {
 			}
 			return SequenceNode.sequence(decls.toArray(new OzNode[decls.size()]), translate(local.body()));
 		} else if (node instanceof CallCommon) {
-			return translateCall((CallCommon) node);
+			return new TailCallCatcherNode(translateCall((CallCommon) node));
```
Comment (eregon) [inline src/org/mozartoz/truffle/translator/Translator.java]: I think this should move in `translateCall`, because all call nodes need to have the catcher.

Comment (eregon) [reply src/org/mozartoz/truffle/translator/Translator.java]: Actually, move it to something like CallNode.create() and review usages of CallNodeGen.create, as they should all be wrapped with a catcher.

File: src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java (hunk 10 above)
Comment (mistasse) [reply src/org/mozartoz/truffle/nodes/call/TailCallThrowerNode.java]: Do you mean the DerefNode or using the Truffle Specialization artillery? (yes, the deref should probably be done manually)

**Comment by eregon** [comment]
Just this execute method is never used, is it?

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
Congrats, works well!

**Pull request merged**